### ISSUE TITLE
event-based strategy for postponing GitLab CI pipelines until `develop` pipelines finish

### ISSUE THREAD

**Issue by zackgalbreath** [spack/spack-infrastructure#342]
![merge_strategy](https://user-images.githubusercontent.com/224720/199295773-b8c96b49-28fe-4d08-99f1-6e4b93c8c165.png)

Here's some pseudo-code laying out our new proposed strategy. Please feel free to propose improvements, clarifications, etc.

* **event: new PR is opened, or existing PR is updated by push**
  * delete any rows in the database for this PR
  * get the currently running develop commit & the latest tested develop commit from GitLab
  * if no currently running develop commit, or `pull.base` is the latest tested develop commit (or one of its ancestors):
    * merge this branch with the latest tested develop commit
    * push this branch to GitLab (triggers CI)
  * else if `pull.base` is the currently running develop commit (or one of its ancestors):
    * merge this branch with the currently running develop commit
    * insert database row (PR_number, to_push=true, to_merge=false)
  * else:
    * insert database row (PR_number, to_push=false, to_merge=true)

* **event: GitLab CI pipeline finishes for develop**
  * for each DB row where 'to_push == true':
    * delete this row from DB
    * push this branch to GitLab (triggers CI)
  * for each DB row where 'to_merge == true':
    * merge PR branch with develop commit whose pipeline just finished 
    * update this row in DB: set to_merge = false, to_push = true

**Comment by alalazo** [comment]
That all seems good to me. The only part where I have a doubt is the edge case:
```
if no currently running develop commit,
```
Can we get, in exceptional circumstances, in a state where:
1. There is no running `develop` commit
2. The current develop is ahead of the latest tested develop commit

? In that case I think we might want to remove that condition.

**Comment by alalazo** [comment]
Another thing that would probably be informative for users is to push a notification status with a message in case they have to wait. Something like:
```
Waiting for the currently running develop pipeline to finish before starting
```
or
```
Waiting for the next develop pipeline to finish before starting
```

**Comment by zackgalbreath** [comment]
Thanks for the feedback, @alalazo 

> Can we get, in exceptional circumstances, in a state where:
> 
> 1. There is no running `develop` commit
> 2. The current develop is ahead of the latest tested develop commit
> 
> ? In that case I think we might want to remove that condition.

For the sake of argument, let's assume that this corner case is a possibility. What should we do if a PR is pushed when there is no currently running develop pipeline, but the most recent develop pipeline is behind HEAD? Options off the top of my head are:

* Merge it with latest-tested-develop and push
* Wait until another develop pipeline starts & finishes

**Comment by alalazo** [comment]
> For the sake of argument, [ ... ]

I think the pseudo-code above would be fine:
1. Merge it with latest-tested-develop and push, if `pull.base` is the latest tested develop or one of the ancestors 
2. Wait until another develop pipeline starts & finishes otherwise

One way in which that might happen is if we can't run, for a given time window, some jobs / services that are needed only when running on develop - but not when running on PRs. Not sure if this is likely, or possible, in reality.

**Comment by zackgalbreath** [comment]
closing for now as this work was superseded by #349
